Add pam service module for user-linked API keys

anodos325 opened the pull request:
This commit adds a dedicated PAM service module for evaluating
user-linked API keys per discussion in NEP-053. Each user may
have one or more keys associated with themselves. Each key is
hashed using pbkdf2-sha512 with a minimum of 300000 iterations.
pbkdf2-sha256 is allowed in some cases for backwards-compatibility
with legacy API tokens.

The module only provides pam_sm_authenticate, other pam
service module functions are stubbed-out.

anodos325 committed:
Add pam service module for user-linked API keys

This commit adds a dedicated PAM service module for evaluating
user-linked API keys per discussion in NEP-053. Each user may
have one or more keys associated with themselves. Each key is
hashed using pbkdf2-sha512 with a minimum of 300000 iterations.

The module only provides pam_sm_authenticate, other pam
service module functions are stubbed-out.

bmeagherix commented on nsswitch/pam_tdb.c:
```diff
@@ -0,0 +1,970 @@
+	}
+
+	/*
+	 * If there were no digitis at all then end == iter_str
```
Typo: digits

bmeagherix commented on nsswitch/pam_tdb.c:
```diff
@@ -0,0 +1,970 @@
+	};
+
+	/*
+	 * generate hash based on password and salt + iterations in pam_tdb entry
```
Seems like the output is base64 encoded (and then munged).  If so, please add a comment to that effect.

bmeagherix commented on nsswitch/pam_tdb.c:
```diff
@@ -0,0 +1,970 @@
+	/* Now use the username to look up password */
+	retval = tdb_auth_request(ctx, username, password, &username_ret);
+	if (username_ret) {
+		pam_set_item (pamh, PAM_USER, username_ret);
```
From https://www.man7.org/linux/man-pages/man3/pam_set_item.3.html
> The pam_set_item function allows applications and PAM service modules to access and to update PAM information of item_type. For this a copy of the object pointed to by the item argument is created.

**If** I'm reading this correctly then we are required to `free(username_ret)` once the above call is done, to avoid a memory leak.

anodos325 replied:
username_ret is allocated via talloc_strdup() and so is automatically freed when we `TALLOC_FREE(ctx)`;

anodos325 committed:
Improve check for key length

anodos325 committed:
Add more comments

bmeagherix commented on nsswitch/pam_tdb.c:
```diff
@@ -0,0 +1,1138 @@
+
+	// First validate that our data is long enough to actually
+	// contain our required fields
+	if ((data + sizeof(int64_t) + sizeof(uint8_t)) > endp) {
```
Should this be:
```
if ((data + sizeof(int64_t) + sizeof(uint32_t) + sizeof(uint8_t)) > endp) {
```
to also take the `entry_db_id` into account?

anodos325 replied:
Yes. Good eye.

anodos325 committed:
Replace strlcpy with memcpy

We just calloced the memcpy target to be explicitly longer than
the string in question.

bmeagherix commented on nsswitch/pam_tdb.c:
```diff
@@ -0,0 +1,1138 @@
+	bytes_read += (sizeof(version) + sizeof(token_cnt));
+	is_admin = strcmp(ctx->admin_user, username) == 0;
+
+	/* We expect password to be in form <db_id>-<key> */
+	if ((key_ptr = strchr(pass, '-')) == NULL) {
+		// malformed key. Force a PAM delay via PAM_AUTH_ERR
+		PAM_CTX_DEBUG(ctx, LOG_ERR,
+			      "malformed password lacks DB ID separator");
+		return PAM_AUTH_ERR;
+	}
```
Huh.  I don't think this format of the password is documented back up the callstack.

Furthermore, do we know that the remote caller knows the DB ID?  Presumably they have to supply <db_id>-<pass>

anodos325 replied:
Yeah, it's part of the string middleware returns when generating the API key.

mgrimesix commented on nsswitch/pam_tdb.c:
```diff
@@ -0,0 +1,1138 @@
+		 * doesn't strip the padding.
+		 */
+		for (i = 0; i < pad_cnt; i++)
+			tmp[strlen(salt_ptr) + i] = '=';
```
Really no need to recalculate strlen each loop.
Actually `strlen(salt_ptr)` could be done once at the start. 🤷‍♂️

anodos325 committed:
Add comment fix boundary check

anodos325 committed:
Address review

anodos325 committed:
Address review

anodos325 committed:
Remove MIN_ITER

We now define this per-algorithm

mgrimesix reviewed (approved):
Nice!
